_examples/basic/5-cqrs-protobuf/main.go

```diff
@@ -207,56 +207,149 @@ func main() {
+		GeneratePublishTopic: func(params cqrs.CommandBusGeneratePublishTopicParams) (string, error) {
 			// we are using queue RabbitMQ config, so we need to have topic per command type
-			return commandName
+			return params.CommandName, nil
 		},
-		CommandHandlers: func(cb *cqrs.CommandBus, eb *cqrs.EventBus) []cqrs.CommandHandler {
-			return []cqrs.CommandHandler{
-				BookRoomHandler{eb},
-				OrderBeerHandler{eb},
-			}
+		OnSend: func(params cqrs.CommandBusOnSendParams) error {
+			logger.Info("Sending command", watermill.LogFields{
+				"command_name": params.CommandName,
+			})
+
+			params.Message.Metadata.Set("sent_at", time.Now().String())
+
+			return nil
 		},
-		CommandsPublisher: commandsPublisher,
-		CommandsSubscriberConstructor: func(handlerName string) (message.Subscriber, error) {
+		Marshaler: cqrsMarshaler,
+		Logger:    logger,
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	commandProcessor, err := cqrs.NewCommandProcessorWithConfig(cqrs.CommandProcessorConfig{
+		GenerateSubscribeTopic: func(params cqrs.CommandProcessorGenerateSubscribeTopicParams) (string, error) {
+			// we are using queue RabbitMQ config, so we need to have topic per command type
+			return params.CommandName, nil
+		},
+		SubscriberConstructor: func(params cqrs.CommandProcessorSubscriberConstructorParams) (message.Subscriber, error) {
 			// we can reuse subscriber, because all commands have separated topics
 			return commandsSubscriber, nil
 		},
-		GenerateEventsTopic: func(eventName string) string {
+		OnHandle: func(params cqrs.CommandProcessorOnHandleParams) error {
+			start := time.Now()
+
+			err := params.Handler.Handle(params.Message.Context(), params.Command)
+
+			logger.Info("Command handled", watermill.LogFields{
+				"command_name": params.CommandName,
+				"duration":     time.Since(start),
+				"err":          err,
+			})
+
+			return err
+		},
+		Marshaler: cqrsMarshaler,
+		Logger:    logger,
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	eventBus, err := cqrs.NewEventBusWithConfig(eventsPublisher, cqrs.EventBusConfig{
+		GeneratePublishTopic: func(params cqrs.GenerateEventPublishTopicParams) (string, error) {
 			// because we are using PubSub RabbitMQ config, we can use one topic for all events
-			return "events"
+			return "events", nil
 
 			// we can also use topic per event type
-			// return eventName
+			// return params.EventName, nil
+		},
+
+		OnPublish: func(params cqrs.OnEventSendParams) error {
+			logger.Info("Publishing event", watermill.LogFields{
+				"event_name": params.EventName,
+			})
+
+			params.Message.Metadata.Set("published_at", time.Now().String())
+
+			return nil
 		},
-		EventHandlers: func(cb *cqrs.CommandBus, eb *cqrs.EventBus) []cqrs.EventHandler {
-			return []cqrs.EventHandler{
-				OrderBeerOnRoomBooked{cb},
-				NewBookingsFinancialReport(),
-			}
+
+		Marshaler: cqrsMarshaler,
+		Logger:    logger,
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	commandProcessor.AddHandler(
```

Comment:
Perhaps it would be a bit more readable to move this after creating the `eventProcessor`, to logically split the constructors from the follow-up method calls.

Reply:
done 👍